File: src/models/Transaction.ts
export interface Transaction {
  id: number;
  date: string;
  payee: string;
  amount: string;
  currency: string;
  plaid_account_name?: string;
  plaidAccountName?: string;
  asset_name?: string;
  assetName?: string;
  account_display_name?: string;
  exclude_from_budget?: boolean;
  excludeFromBudget?: boolean;
  exclude_from_totals?: boolean;
  excludeFromTotals?: boolean;
  is_group?: boolean;
  isGroup?: boolean;
  group_id?: number;
  groupId?: number;
  category_name?: string;
  categoryName?: string;
}

export class TransactionUtils {
  static getAmount(transaction: Transaction): number {
    const cleanAmount = transaction.amount.replace(/[^0-9.-]/g, '');
    return parseFloat(cleanAmount) || 0;
  }

  static getAccountName(transaction: Transaction): string | null {
    // Try all possible field names from the API
    const accountName =
      transaction.plaid_account_name ||
      transaction.plaidAccountName ||
      transaction.asset_name ||
      transaction.assetName ||
      transaction.account_display_name;

    if (!accountName) return null;

    // Return raw account name, truncated to 10 characters
    if (accountName.length > 10) {
      return accountName.substring(0, 10);
    }

    return accountName;
  }

  static getFormattedDate(transaction: Transaction): string {
    const date = new Date(transaction.date);
    const options: Intl.DateTimeFormatOptions = {
      weekday: 'short',
      month: 'short',
      day: 'numeric'
    };
    return date.toLocaleDateString('en-US', options);
  }
}

File: PerchIOS/PerchIOS/PerchIOS/PerchIOS/Models/Transaction.swift
import Foundation

struct Transaction: Identifiable, Codable {
    let id: Int
    let date: String
    let payee: String
    let amount: String
    let currency: String

    // Various field names from API
    let plaidAccountName: String?
    let plaid_account_name: String?
    let assetName: String?
    let asset_name: String?
    let account_display_name: String?

    let excludeFromBudget: Bool?
    let exclude_from_budget: Bool?
    let excludeFromTotals: Bool?
    let exclude_from_totals: Bool?
    let isGroup: Bool?
    let is_group: Bool?
    let groupId: Int?
    let group_id: Int?
    let categoryName: String?
    let category_name: String?

    var displayAmount: Double {
        let cleanAmount = amount.replacingOccurrences(of: "[^0-9.-]", with: "", options: .regularExpression)
        return Double(cleanAmount) ?? 0
    }

    var accountName: String? {
        // Try all possible field names from the API
        if let name = plaidAccountName ?? plaid_account_name ?? assetName ?? asset_name ?? account_display_name {
            // Return raw account name, truncated to 10 characters
            return name.count > 10 ? String(name.prefix(10)) : name
        }
        return nil
    }

    var formattedDate: String {
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd"
        guard let date = dateFormatter.date(from: date) else { return date }

        dateFormatter.dateFormat = "MMM d, yyyy"
        return dateFormatter.string(from: date)
    }

    var shouldExclude: Bool {
        return excludeFromTotals ?? exclude_from_totals ?? false
    }

    var category: String? {
        return categoryName ?? category_name
    }
}

class TransactionState: ObservableObject {
    @Published var transaction: Transaction

    init(_ transaction: Transaction) {
        self.transaction = transaction
    }
}